### PR TITLE
Add nightly turtlebot

### PR DESCRIPTION
**Pull request by clalancette**
The change I really wanted to do was to add a nightly build job for turtlebot.  However, I found the way the current jobs are created to be confusing, so I also rewrote the loop in the first patch.  See the commit message for the first patch for an explanation.

I tested this in a dry-run locally; the only things that were going to be different from before were a small typo fix in the nightly_*_repeated jobs (an extra space at the end of the spec), and the creation of the two new nightly turtlebot jobs.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Please double check the current configuration by comparing the newly generated config with the existing jobs (without actually overwriting them). You will notice that the patch has a couple of bugs atm.

**Comment by clalancette** [comment]
Ah, good call.  I missed that.

File: create_jenkins_job.py [hunk 1]
```diff
@@ -114,73 +116,87 @@ def main(argv=None):
...
+        })
+
+        # configure turtlebot jobs on Linux only for now
+        if os_name in ['linux', 'linux-aarch64']:
+            create_job(os_name, 'ci_turtlebot-demo_%s' % (os_name), 'ci_job.xml.em', {
+                'cmake_build_type': 'None',
+                'turtlebot_demo': True,
+                'supplemental_repos_url': 'https://raw.githubusercontent.com/ros2/turtlebot2_demo/master/turtlebot2_demo.repos',
+            })
+            create_job(os_name, 'nightly_turtlebot-demo-%s' % (os_name), 'ci_job.xml.em', {
```
Comment (dhood) [inline create_jenkins_job.py]: this needs to set turtlebot_demo and supplemental_repos_url, right? I'd also recommend the name include debug for consistency

Comment (mikaelarguedas) [inline create_jenkins_job.py]: `'turtlebot_demo': True,` << I think that this is needed as well

Comment (clalancette) [inline create_jenkins_job.py]: Yep, you are exactly right, totally missed.  Thanks, I've added that now, along with the "debug" name.  Please take another look when you have a moment.

File: create_jenkins_job.py [hunk 2]
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
...
         # configure nightly triggered job
-        job_name = 'nightly_' + os_name + '_debug'
-        if os_name == 'win':
-            job_name = job_name[0:-2]
-        job_data['cmake_build_type'] = 'Debug'
-        job_config = expand_template('ci_job.xml.em', job_data)
-        configure_job(jenkins, job_name, job_config, **jenkins_kwargs)
+        short_name = 'nightly_' + os_name + '_debug'
+        if os_name == 'windows':
+            short_name = 'nightly_win_deb'
```
Comment (mikaelarguedas) [inline create_jenkins_job.py]: Is it necessary to change the logic of how the names shorten for windows ? (same everywhere the job names are hardcoded)

File: create_jenkins_job.py [hunk 3]
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
...
 
         # configure nightly coverage job on x86 Linux only
         if os_name == 'linux':
-            job_name = 'nightly_' + os_name + '_coverage'
-            job_data['cmake_build_type'] = 'Debug'
-            job_data['enable_c_coverage_default'] = 'true'
-            job_config = expand_template('ci_job.xml.em', job_data)
-            configure_job(jenkins, job_name, job_config, **jenkins_kwargs)
-            job_data['enable_c_coverage_default'] = 'false'
+            create_job(os_name, 'nightly_linux_coverage', 'ci_job.xml.em', {
```
Comment (mikaelarguedas) [inline create_jenkins_job.py]: Same remark asabove, I'm not sure hardcoding the job names is a good idea. If we want to create a coverage job for osx we will need to create another conditional bloc while we could just us this one if it stayed as it was before

File: create_jenkins_job.py [hunk 4]
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
...
+
+        # configure turtlebot jobs on Linux only for now
+        if os_name in ['linux', 'linux-aarch64']:
+            create_job(os_name, 'ci_turtlebot-demo_%s' % (os_name), 'ci_job.xml.em', {
+                'cmake_build_type': 'None',
+                'turtlebot_demo': True,
+                'supplemental_repos_url': 'https://raw.githubusercontent.com/ros2/turtlebot2_demo/master/turtlebot2_demo.repos',
+            })
+            create_job(os_name, 'nightly_turtlebot-demo_%s_debug' % (os_name), 'ci_job.xml.em', {
+                'cmake_build_type': 'Debug',
```
Comment (mikaelarguedas) [inline create_jenkins_job.py]: Note: Cartographer doesn't build in `Debug` AFAIK so I would expect this job to fail every night

File: create_jenkins_job.py [hunk 5]
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
...
+            'cmake_build_type': 'None',
+            'time_trigger_spec': PERIODIC_JOB_SPEC,
+            'mailer_recipients': DEFAULT_MAIL_RECIPIENTS,
+            'ament_test_args_default': '--retest-until-fail 20 --ctest-args -LE linter --',
+        })
+
+        # configure turtlebot jobs on Linux only for now
+        if os_name in ['linux', 'linux-aarch64']:
+            create_job(os_name, 'ci_turtlebot-demo_%s' % (os_name), 'ci_job.xml.em', {
+                'cmake_build_type': 'None',
```
Comment (mikaelarguedas) [inline create_jenkins_job.py]: We usually never have None nightlies, is there a reason to have one for the turtlebot job ?

File: create_jenkins_job.py [hunk 6]
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
...
+        create_job(os_name, short_name, 'ci_job.xml.em', {
+            'cmake_build_type': 'None',
+            'time_trigger_spec': PERIODIC_JOB_SPEC,
+            'mailer_recipients': DEFAULT_MAIL_RECIPIENTS,
+            'ament_test_args_default': '--retest-until-fail 20 --ctest-args -LE linter --',
+        })
+
+        # configure turtlebot jobs on Linux only for now
+        if os_name in ['linux', 'linux-aarch64']:
+            create_job(os_name, 'ci_turtlebot-demo_%s' % (os_name), 'ci_job.xml.em', {
```
Comment (mikaelarguedas) [inline create_jenkins_job.py]: every other job uses a `job_name` variable and uses `+` for concatenation, can you do the same here to stay consistent?

File: create_jenkins_job.py (hunk 3 above)
Comment (clalancette) [inline create_jenkins_job.py]: OK, yeah.  On this one, I agree with you.  I'll change this back to building it up from the os_name.

File: create_jenkins_job.py (hunk 6 above)
Comment (clalancette) [inline create_jenkins_job.py]: Yeah, that's fine, I'll fix that.  This was a cut-n-paste from below, but I agree with you, it should be consistent.

File: create_jenkins_job.py (hunk 5 above)
Comment (clalancette) [inline create_jenkins_job.py]: This was a cut-n-paste from the old turtlebot job.  I'm not sure where that configuration originally came from.

File: create_jenkins_job.py (hunk 4 above)
Comment (clalancette) [inline create_jenkins_job.py]: Ah, I had forgotten about that fact, good call.  Will change it to release instead.

File: create_jenkins_job.py (hunk 2 above)
Comment (clalancette) [inline create_jenkins_job.py]: Not really necessary, no, but the previous way seemed overly complicated to me.  It was, in effect, hard-coding the short names, just in a less obvious way.

**Comment by clalancette** [comment]
> If the goal is to confirm that the code still builds I'm not sure we need to create several build types because it will take a lot of buildfarm time.

Sorry, I missed this bit before.  As it stands, this patch will create 2 new jobs: nightly_turtlebot-demo_linux and nightly_turtlebot-demo_linux-aarch64.  Are you worried that adding those two nightly builds will tax the build system, or were you concerned about something else?

**Comment by mikaelarguedas** [comment]
> Sorry, I missed this bit before. As it stands, this patch will create 2 new jobs: nightly_turtlebot-demo_linux and nightly_turtlebot-demo_linux-aarch64. Are you worried that adding those two nightly builds will tax the build system, or were you concerned about something else?

I was concerned about mutliple turtlebot jobs **per architecture** (e.g. before one Debug and one None). If the current patch "create 2 new jobs: nightly_turtlebot-demo_linux and nightly_turtlebot-demo_linux-aarch64" that's good :+1:

File: create_jenkins_job.py [hunk 7]
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
...
+        })
+
+        # configure turtlebot jobs on Linux only for now
+        if os_name in ['linux', 'linux-aarch64']:
+            create_job(os_name, 'ci_turtlebot-demo_' + os_name, 'ci_job.xml.em', {
+                'cmake_build_type': 'None',
+                'turtlebot_demo': True,
+                'supplemental_repos_url': 'https://raw.githubusercontent.com/ros2/turtlebot2_demo/master/turtlebot2_demo.repos',
+            })
+            create_job(os_name, 'nightly_turtlebot-demo_' + os_name, 'ci_job.xml.em', {
```
Comment (mikaelarguedas) [inline create_jenkins_job.py]: > I'd also recommend the name include debug for consistency

This still applies, we should add `_release` to the job name here

File: create_jenkins_job.py (hunk 2 above)
Comment (mikaelarguedas) [inline create_jenkins_job.py]: the previous logic was creating the job_name the same way for everyone and then modifying it to make it shorter for windows. This new approach replace the variable content with a hard-coded string literal discarding the previous content so I think it's more likely that the naming becomes inconsistent if we change one but not the other

**Comment by clalancette** [comment]
All right, I added release back to the name, and I've changed it up so that we are building the Windows names again.  I still made some changes in how we built the names; in particular I made it so that they have a common set of rules.  Ideally we'd apply those rules to all jobs, but that would currently replace some of the windows jobs with other jobs (ci_windows -> ci_win, packaging_windows -> packaging_win, etc).  While that is technically more consistent, I don't think we need to do that right now, so I just special cased the ones that were special-cased before.  @mikaelarguedas please take another look when you get a chance.

**Comment by mikaelarguedas** [comment]
> I made it so that they have a common set of rules. Ideally we'd apply those rules to all jobs,

:+1: agreed that a common rule makes more sense but appreciate that you apply them only to the jobs that needed it to keep the naming. Ideally the rule of cropping after the 15th character should be shared as well rather than duplicated for all the jobs needing cropping

**Comment by clalancette** [comment]
Thanks for the review.  I'm going to merge this and then deploy it.

**Pull request merged**